{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3238afc1",
   "metadata": {},
   "source": [
    "# Create a VCF from the genotypes\n",
    "We simulated the parent-offspring genotypes (hold the applause), and now need to somehow make a VCF file out of it as input for `Mimick`. The genotypes are given as `0` and `1`, so we need to parse the human reference genome to find the `0` (aka 'Ref') and make up `1` (aka 'Alt'). After that, we need to encode it into a VCF file, something simple enough that it satisfies what Mimick needs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1631e349",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import pysam\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8409b489",
   "metadata": {},
   "source": [
    "## Download the human genome `GRCh38.p14`\n",
    "We need the human genome that was used for the pedigree genotypes (contig names, lengths, etc.) We'll download the entire thing, then subset it to keep only the first 5 autosomes that we are interested in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebfd2b15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "972898531"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRCh38 = \"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_genomic.fna.gz\"\n",
    "os.makedirs(\"../reference_assembly\", exist_ok=True)\n",
    "assembly_original = os.path.join(\"../reference_assembly\", \"GRCh38.p14.fasta.gz\")\n",
    "assembly_trunc = os.path.join(\"../reference_assembly\", \"GRCh38.p14.5A.fa\")\n",
    "r = requests.get(GRCh38)\n",
    "open(assembly_original, 'wb').write(r.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "292f3698",
   "metadata": {},
   "source": [
    "Here we are subsetting `GRCh38` for the 5 autosomes we are interested in. We'll also use this loop to calculate the length of each chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "06c4e3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_chrom = {\"NC_000001.11\" : 0, \"NC_000002.12\" : 0, \"NC_000003.12\" : 0, \"NC_000004.12\" : 0, \"NC_000005.10\" : 0}\n",
    "with (\n",
    "    gzip.open(assembly_original, 'r') as fain,\n",
    "    open(assembly_trunc, 'w') as faout,\n",
    "    ):\n",
    "    write = False\n",
    "    _name = \"\"\n",
    "    for i in fain:\n",
    "        if i.startswith(b\">\"):\n",
    "            _name = i.decode(\"utf8\").split()[0].lstrip(\">\")\n",
    "            if _name in target_chrom:\n",
    "                faout.write(f\">{_name}\\n\")\n",
    "                _write = True\n",
    "            else:\n",
    "                _write = False\n",
    "        else:\n",
    "            if _write:\n",
    "                target_chrom[_name] += len(i)\n",
    "                faout.write(i.decode(\"utf-8\"))\n",
    "\n",
    "os.system(f\"bgzip {assembly_trunc}\")\n",
    "assembly_trunc += \".gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7e71495e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NC_000001.11': 252068378,\n",
       " 'NC_000002.12': 245220949,\n",
       " 'NC_000003.12': 200774254,\n",
       " 'NC_000004.12': 192592237,\n",
       " 'NC_000005.10': 183807488}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_chrom"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a52b58e",
   "metadata": {},
   "source": [
    "## Start with a template\n",
    "How do we know what will work for Mimick? Great question. Even though I wrote the thing, I don't offhand remember what fields the VCF requires. I could check, but where's the fun in that. Instead, let's just pull the VCF file Mimick uses for testing and explore that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d9f54cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://github.com/pdimens/mimick/raw/refs/heads/main/test/test.vcf.gz'\n",
    "r = requests.get(url)\n",
    "open(\"test.vcf.gz\" , 'wb').write(r.content)\n",
    "vcf_template = [i.decode(\"utf-8\").strip() for i in gzip.open(\"test.vcf.gz\" , 'r').readlines()]\n",
    "vcf_header = [i for i in vcf_template if i.startswith(\"#\")]\n",
    "print(\"\\n\".join(vcf_header))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e770ef5e",
   "metadata": {},
   "source": [
    "## Update the header\n",
    "Let's update the header to:\n",
    "1. remove entries that include `bcftools` and add our parent and offspring names to the column names\n",
    "2. for the sake of being pedantic, let's also replace the `source` so it's clear where this file is from (us, here, doing this)\n",
    "3. replace the `contig` part with the list of the actual 5 contigs we are working with and their lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "49d99cc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##source=PhasingAssemblySims - Pavel Dimens\n",
      "##contig=<ID=NC_000001.11,length=252068378>\n",
      "##contig=<ID=NC_000002.12,length=245220949>\n",
      "##contig=<ID=NC_000003.12,length=200774254>\n",
      "##contig=<ID=NC_000004.12,length=192592237>\n",
      "##contig=<ID=NC_000005.10,length=183807488>\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tfather\tmother\toffspring_1\toffspring_2\toffspring_3\toffspring_4\toffspring_5\toffspring_6\toffspring_7\toffspring_8\toffspring_9\toffspring_10\n"
     ]
    }
   ],
   "source": [
    "vcf_new_header = [i for i in vcf_header if not i.startswith(\"##bcftools\")]\n",
    "\n",
    "# replace sample names\n",
    "sample_names = [\"father\",\"mother\"] + [f\"offspring_{i}\" for i in range(1,11)]\n",
    "vcf_new_header[-1] = '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t' + \"\\t\".join(sample_names)\n",
    "vcf_new_header[2] = \"##source=PhasingAssemblySims - Pavel Dimens\"\n",
    "\n",
    "# update contig names/lengths\n",
    "idx = 3\n",
    "del vcf_new_header[idx]\n",
    "for k,v in target_chrom.items():\n",
    "    vcf_new_header.insert(idx, f\"##contig=<ID={k},length={v}>\")\n",
    "    idx += 1\n",
    "\n",
    "print(\"\\n\".join(vcf_new_header))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e080bfea",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"pedigree.vcf\", \"w\") as _vcf:\n",
    "    _vcf.write(\"\\n\".join(vcf_new_header))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "408820f9",
   "metadata": {},
   "source": [
    "## `CHROM`, `POS`, `REF` and `ALT`\n",
    "We need to populate the CHROM, POS, REF, and ALT columns, but since this is just a text file (and not a table), we should probably do this line-by-line. To achieve that, we'll need to read in the pedigree genotype matrix we created. That way, we can go row-by-row in the pedigree and:\n",
    "1. get the chromosome and position with which to index the genome\n",
    "    - we'll use `pysam` for random access\n",
    "2. find the nucleotide at that position and choose \"any other base\"\n",
    "    - REF A/T -> ALT G/C\n",
    "    - REF G/C -> ALT A/T\n",
    "    - **will have to manually inspect for `N`s**\n",
    "3. collapse the sample genotypes into the `GT` field as-is (since they are already in `0`|`1` format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "3e71cacf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# creata a fasta index for random access\n",
    "os.system(f\"samtools faidx {assembly_trunc}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
