{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3238afc1",
   "metadata": {},
   "source": [
    "# Format a templated VCF\n",
    "We simulated the parent-offspring genotypes (hold the applause), and now need to somehow make a VCF file out of it as input for `Mimick`. The genotypes are given as `0` and `1`, so we need to parse the human reference genome to find the `0` (aka 'Ref') and make up `1` (aka 'Alt'). After that, we need to encode it into a VCF file, something simple enough that it satisfies what Mimick needs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1631e349",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import pysam\n",
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e71495e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NC_000001.11': 248956422,\n",
       " 'NC_000002.12': 242193529,\n",
       " 'NC_000003.12': 198295559,\n",
       " 'NC_000004.12': 190214555,\n",
       " 'NC_000005.10': 181538259}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_chrom = {}\n",
    "with open(\"../reference_assembly/GRCh38.p14.5A.fa.gz.fai\", 'r') as fai:\n",
    "    for line in fai:\n",
    "        _line = line.split()\n",
    "        target_chrom[_line[0]] = int(_line[1])\n",
    "\n",
    "target_chrom"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a52b58e",
   "metadata": {},
   "source": [
    "## Start with a template\n",
    "How do we know what will work for Mimick? Great question! Even though I wrote the thing, I don't offhand remember what fields the VCF requires. I could check, but where's the fun in that. Instead, let's just pull the VCF file Mimick uses for testing and explore that. It already passes Mimick's tests, so surely it should work here. ðŸ¤·"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7d9f54cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##source=tskit 0.5.6\n",
      "##contig=<ID=chr1,length=1500001>\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "##bcftools_viewVersion=1.22+htslib-1.22\n",
      "##bcftools_viewCommand=view -Oz -s tsk_0,tsk_1,tsk_2,tsk_3,tsk_4,tsk_5,tsk_6,tsk_7,tsk_8,tsk_9,tsk_10,tsk_11,tsk_12,tsk_13,tsk_14,tsk_15,tsk_16,tsk_17,tsk_18,tsk_19,tsk_20,tsk_21,tsk_22,tsk_23,tsk_24,tsk_25,tsk_26,tsk_27,tsk_28,tsk_29,tsk_30,tsk_31,tsk_32,tsk_33,tsk_34,tsk_35,tsk_36,tsk_37,tsk_38,tsk_39,tsk_40,tsk_41,tsk_42,tsk_43,tsk_44,tsk_45,tsk_46,tsk_47,tsk_48,tsk_49,tsk_50,tsk_51,tsk_52,tsk_53,tsk_54,tsk_55,tsk_56,tsk_57,tsk_58,tsk_59,tsk_60,tsk_61,tsk_62,tsk_63 genos.vcf; Date=Fri Jul 11 07:23:04 2025\n",
      "##bcftools_viewCommand=view data/example-vcf-for-pavel.vcf.gz; Date=Mon Jul 14 10:26:31 2025\n",
      "##bcftools_viewCommand=view -Oz; Date=Mon Jul 14 10:26:31 2025\n",
      "##bcftools_viewCommand=view new-vcf-for-pavel.vcf.gz; Date=Mon Jul 14 15:27:24 2025\n",
      "##bcftools_viewVersion=1.22+htslib-1.22.1\n",
      "##bcftools_viewCommand=view -Oz test.vcf; Date=Thu Jul 17 11:47:27 2025\n",
      "##bcftools_viewCommand=view -s tsk_0,tsk_1,tsk_2,tsk_3,tsk_4,tsk_5,tsk_6 -Oz test.vcf.gz; Date=Mon Jul 21 09:52:05 2025\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\ttsk_0\ttsk_1\ttsk_2\ttsk_3\ttsk_4\ttsk_5\ttsk_6\n"
     ]
    }
   ],
   "source": [
    "url = 'https://github.com/pdimens/mimick/raw/refs/heads/main/test/test.vcf.gz'\n",
    "r = requests.get(url)\n",
    "open(\"test.vcf.gz\" , 'wb').write(r.content)\n",
    "vcf_template = [i.decode(\"utf-8\").strip() for i in gzip.open(\"test.vcf.gz\" , 'r').readlines()]\n",
    "vcf_header = [i for i in vcf_template if i.startswith(\"#\")]\n",
    "print(\"\\n\".join(vcf_header))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e770ef5e",
   "metadata": {},
   "source": [
    "## Update the header\n",
    "Let's update the header to:\n",
    "1. remove entries that include `bcftools` and add our parent and offspring names to the column names\n",
    "2. for the sake of being pedantic, let's also replace the `source` so it's clear where this file is from (us, here, doing this)\n",
    "3. replace the `contig` part with the list of the actual 5 contigs we are working with and their lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "49d99cc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.2\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
      "##source=PhasingAssemblySims - Pavel Dimens\n",
      "##contig=<ID=NC_000001.11,length=248956422>\n",
      "##contig=<ID=NC_000002.12,length=242193529>\n",
      "##contig=<ID=NC_000003.12,length=198295559>\n",
      "##contig=<ID=NC_000004.12,length=190214555>\n",
      "##contig=<ID=NC_000005.10,length=181538259>\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tfather\tmother\toffspring_1\toffspring_2\toffspring_3\toffspring_4\toffspring_5\toffspring_6\toffspring_7\toffspring_8\toffspring_9\toffspring_10\n"
     ]
    }
   ],
   "source": [
    "vcf_new_header = [i for i in vcf_header if not i.startswith(\"##bcftools\")]\n",
    "\n",
    "# replace sample names\n",
    "sample_names = [\"father\",\"mother\"] + [f\"offspring_{i}\" for i in range(1,11)]\n",
    "vcf_new_header[-1] = '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t' + \"\\t\".join(sample_names)\n",
    "vcf_new_header[2] = \"##source=PhasingAssemblySims - Pavel Dimens\"\n",
    "\n",
    "# update contig names/lengths\n",
    "idx = 3\n",
    "del vcf_new_header[idx]\n",
    "for k,v in target_chrom.items():\n",
    "    vcf_new_header.insert(idx, f\"##contig=<ID={k},length={v}>\")\n",
    "    idx += 1\n",
    "\n",
    "print(\"\\n\".join(vcf_new_header))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e080bfea",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"pedigree.vcf\", \"w\") as _vcf:\n",
    "    _vcf.write(\"\\n\".join(vcf_new_header) + \"\\n\")\n",
    "\n",
    "# remove the template VCF\n",
    "os.remove(\"test.vcf.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e861240",
   "metadata": {},
   "source": [
    "The REF/ALT assignments are handled in the [reference curation](../reference_assembly/setup_reference.ipynb) and populating the records will be handled in the [pedigree simulation](./simulate_pedigree.ipynb)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
