{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3238afc1",
   "metadata": {},
   "source": [
    "# Simulate Pedigree Genotypes\n",
    "This notebook runs through creating the genotypes/haplotypes for 2  parents and their 10 offspring. The intent is to create the genotypes and convert to a VCF as input into `Mimick`, which will then generate linked-read data for phasing and assembly simulations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d265f65a",
   "metadata": {
    "tags": [
     "hide-cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#| install missing pacakges\n",
    "list.of.packages <- c(\"dplyr\", \"simfam\", \"tibble\", \"tidyr\")\n",
    "new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,\"Package\"])]\n",
    "if(length(new.packages)) install.packages(new.packages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "736334cf",
   "metadata": {
    "tags": [
     "remove-output"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(simfam)\n",
    "library(tibble)\n",
    "library(tidyr)\n",
    "set.seed(6969)\n",
    "genbank <- FALSE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2e5f18c",
   "metadata": {},
   "source": [
    "## Simulate Genotypes with `simfam`\n",
    "The code in the next few cells is taken from the [simfam](https://cran.r-project.org/web/packages/simfam/refman/simfam.html#simfam-package) R package documentation for the `recomb_geno_inds()` function and modified to simulate 10 individuals using the first 5 human autosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e6706da",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 12 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>id</th><th scope=col>pat</th><th scope=col>mat</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>father      </td><td>NA    </td><td>NA    </td></tr>\n",
       "\t<tr><td>mother      </td><td>NA    </td><td>NA    </td></tr>\n",
       "\t<tr><td>offspring_1 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_2 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_3 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_4 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_5 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_6 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_7 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_8 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_9 </td><td>father</td><td>mother</td></tr>\n",
       "\t<tr><td>offspring_10</td><td>father</td><td>mother</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 12 × 3\n",
       "\\begin{tabular}{lll}\n",
       " id & pat & mat\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t father       & NA     & NA    \\\\\n",
       "\t mother       & NA     & NA    \\\\\n",
       "\t offspring\\_1  & father & mother\\\\\n",
       "\t offspring\\_2  & father & mother\\\\\n",
       "\t offspring\\_3  & father & mother\\\\\n",
       "\t offspring\\_4  & father & mother\\\\\n",
       "\t offspring\\_5  & father & mother\\\\\n",
       "\t offspring\\_6  & father & mother\\\\\n",
       "\t offspring\\_7  & father & mother\\\\\n",
       "\t offspring\\_8  & father & mother\\\\\n",
       "\t offspring\\_9  & father & mother\\\\\n",
       "\t offspring\\_10 & father & mother\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 12 × 3\n",
       "\n",
       "| id &lt;chr&gt; | pat &lt;chr&gt; | mat &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| father       | NA     | NA     |\n",
       "| mother       | NA     | NA     |\n",
       "| offspring_1  | father | mother |\n",
       "| offspring_2  | father | mother |\n",
       "| offspring_3  | father | mother |\n",
       "| offspring_4  | father | mother |\n",
       "| offspring_5  | father | mother |\n",
       "| offspring_6  | father | mother |\n",
       "| offspring_7  | father | mother |\n",
       "| offspring_8  | father | mother |\n",
       "| offspring_9  | father | mother |\n",
       "| offspring_10 | father | mother |\n",
       "\n"
      ],
      "text/plain": [
       "   id           pat    mat   \n",
       "1  father       NA     NA    \n",
       "2  mother       NA     NA    \n",
       "3  offspring_1  father mother\n",
       "4  offspring_2  father mother\n",
       "5  offspring_3  father mother\n",
       "6  offspring_4  father mother\n",
       "7  offspring_5  father mother\n",
       "8  offspring_6  father mother\n",
       "9  offspring_7  father mother\n",
       "10 offspring_8  father mother\n",
       "11 offspring_9  father mother\n",
       "12 offspring_10 father mother"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# The pedigree (a PLINK style fam table)\n",
    "fam <- tibble(\n",
    "  id = c(c(\"father\", \"mother\"), paste0(\"offspring_\", 1:10)),\n",
    "  pat = c(NA, NA, rep(\"father\", 10)),\n",
    "  mat = c(NA, NA, rep(\"mother\", 10))\n",
    ")\n",
    "fam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b70aa0e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "Use the first 5 chromosomes of the latest human recombination map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d1ef6f4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pos</th><th scope=col>posg</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>      1</td><td>0.0000000</td></tr>\n",
       "\t<tr><td> 285245</td><td>0.5601822</td></tr>\n",
       "\t<tr><td>1026830</td><td>2.7201562</td></tr>\n",
       "\t<tr><td>1779975</td><td>3.6594862</td></tr>\n",
       "\t<tr><td>1942086</td><td>3.6880442</td></tr>\n",
       "\t<tr><td>2049679</td><td>4.4486302</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 2\n",
       "\\begin{tabular}{ll}\n",
       " pos & posg\\\\\n",
       " <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t       1 & 0.0000000\\\\\n",
       "\t  285245 & 0.5601822\\\\\n",
       "\t 1026830 & 2.7201562\\\\\n",
       "\t 1779975 & 3.6594862\\\\\n",
       "\t 1942086 & 3.6880442\\\\\n",
       "\t 2049679 & 4.4486302\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 2\n",
       "\n",
       "| pos &lt;int&gt; | posg &lt;dbl&gt; |\n",
       "|---|---|\n",
       "|       1 | 0.0000000 |\n",
       "|  285245 | 0.5601822 |\n",
       "| 1026830 | 2.7201562 |\n",
       "| 1779975 | 3.6594862 |\n",
       "| 1942086 | 3.6880442 |\n",
       "| 2049679 | 4.4486302 |\n",
       "\n"
      ],
      "text/plain": [
       "  pos     posg     \n",
       "1       1 0.0000000\n",
       "2  285245 0.5601822\n",
       "3 1026830 2.7201562\n",
       "4 1779975 3.6594862\n",
       "5 1942086 3.6880442\n",
       "6 2049679 4.4486302"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "map <- recomb_map_hg38[1L:5L]\n",
    "head(map[[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cac6c9a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "Initialize the parents (founders)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9e27eac1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "founders <- recomb_init_founders(c(\"father\", \"mother\"), map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bde48c11",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "Draw the recombination breaks for the children and add base pair coordinates to the recombination breaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7de6ae57",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "inds <- recomb_fam(founders, fam)\n",
    "\n",
    "inds <- recomb_map_inds(inds, map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66502bdb",
   "metadata": {},
   "source": [
    "Create random ancestral haplotypes. Based on the average of 1 SNP per 1300bp ([Miller et al, 2007](https://pmc.ncbi.nlm.nih.gov/articles/PMC1885222/)) reported for hg_38, we can back-of-the-envelope calculate the number of SNPs for the first 5 chromosomes:\n",
    "\n",
    "| chromosome | length (bp) | number of SNPs (length / 1300) |\n",
    "|:-----------|------------:|-------------------------------:|\n",
    "| 1 | 248,956,422 | 191,505 |\n",
    "| 2 | 242,193,529 | 186,303 |\n",
    "| 3 | 198,295,559 | 152,535 |\n",
    "| 4 | 190,214,555 | 146,319 |\n",
    "| 5 | 181,538,259 | 139,645 |\n",
    "\n",
    "There are lots of Ns in the human genome, so we need to include loci from only non-N positions, which we derived in [the reference setup](../reference_assembly/setup_reference.ipynb). We'll start by reading in those positions, then the loop will subset the positions for a given chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8f9f182b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>chrom</th><th scope=col>pos</th><th scope=col>id</th><th scope=col>ref</th><th scope=col>alt</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>NC_000001.11</td><td>10870</td><td>1</td><td>G</td><td>C</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>NC_000001.11</td><td>11245</td><td>2</td><td>G</td><td>A</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>NC_000001.11</td><td>13626</td><td>3</td><td>G</td><td>T</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>NC_000001.11</td><td>15195</td><td>4</td><td>T</td><td>G</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NC_000001.11</td><td>15981</td><td>5</td><td>A</td><td>C</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>NC_000001.11</td><td>17245</td><td>6</td><td>T</td><td>A</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & chrom & pos & id & ref & alt\\\\\n",
       "  & <chr> & <int> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & NC\\_000001.11 & 10870 & 1 & G & C\\\\\n",
       "\t2 & NC\\_000001.11 & 11245 & 2 & G & A\\\\\n",
       "\t3 & NC\\_000001.11 & 13626 & 3 & G & T\\\\\n",
       "\t4 & NC\\_000001.11 & 15195 & 4 & T & G\\\\\n",
       "\t5 & NC\\_000001.11 & 15981 & 5 & A & C\\\\\n",
       "\t6 & NC\\_000001.11 & 17245 & 6 & T & A\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | chrom &lt;chr&gt; | pos &lt;int&gt; | id &lt;int&gt; | ref &lt;chr&gt; | alt &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | NC_000001.11 | 10870 | 1 | G | C |\n",
       "| 2 | NC_000001.11 | 11245 | 2 | G | A |\n",
       "| 3 | NC_000001.11 | 13626 | 3 | G | T |\n",
       "| 4 | NC_000001.11 | 15195 | 4 | T | G |\n",
       "| 5 | NC_000001.11 | 15981 | 5 | A | C |\n",
       "| 6 | NC_000001.11 | 17245 | 6 | T | A |\n",
       "\n"
      ],
      "text/plain": [
       "  chrom        pos   id ref alt\n",
       "1 NC_000001.11 10870 1  G   C  \n",
       "2 NC_000001.11 11245 2  G   A  \n",
       "3 NC_000001.11 13626 3  G   T  \n",
       "4 NC_000001.11 15195 4  T   G  \n",
       "5 NC_000001.11 15981 5  A   C  \n",
       "6 NC_000001.11 17245 6  T   A  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_pos <- read.table(\n",
    "    \"../reference_assembly/GRCh38.p14.snp.positions\",\n",
    "    col.names = c(\"chrom\", \"pos\", \"id\", \"ref\", \"alt\"),\n",
    "    header = F\n",
    " )\n",
    "chr_names <- unique(snp_pos$chrom)\n",
    "\n",
    "head(snp_pos)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e6aa370",
   "metadata": {},
   "source": [
    "Now to loop through the chromosomes and simulate haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4a4d68f2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "haplo <- vector( 'list', length( map ) )\n",
    "\n",
    "# names of ancestor haplotypes for this scenario\n",
    "# (founders of fam$id but each with \"_pat\" and \"_mat\" suffixes)\n",
    "anc_names <- c('father_pat', 'father_mat', 'mother_pat', 'mother_mat')\n",
    "n_ind <- length(anc_names)\n",
    "\n",
    "idx <- 0\n",
    "for (chr in seq_along(map)) {\n",
    "    idx <- idx + 1\n",
    "    # get snp positions\n",
    "    .chrom <- chr_names[idx]\n",
    "    pos_chr <- snp_pos[snp_pos$chrom == .chrom, \"pos\"]\n",
    "    n_loci <- length(pos_chr)\n",
    "    # draw haplotypes\n",
    "    X_chr <- matrix(\n",
    "        rbinom(n_loci * n_ind, 1L, 0.5),\n",
    "        nrow = n_loci,\n",
    "        ncol = n_ind\n",
    "    )\n",
    "    # required column names!\n",
    "    colnames(X_chr) <- anc_names\n",
    "    # add to structure, in a list\n",
    "    haplo[[chr]] <- list(X = X_chr, pos = pos_chr)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3ab9a8b1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ father:List of 2\n",
      "  ..$ pat:List of 5\n",
      "  .. ..$ : int [1:177293] 1 1 0 0 0 1 1 1 0 0 ...\n",
      "  .. ..$ : int [1:185037] 1 0 0 1 0 0 1 0 0 1 ...\n",
      "  .. ..$ : int [1:152385] 1 0 1 0 1 0 1 1 1 0 ...\n",
      "  .. ..$ : int [1:145964] 0 1 0 1 1 0 1 1 0 0 ...\n",
      "  .. ..$ : int [1:139435] 1 1 1 1 1 1 0 0 1 0 ...\n",
      "  ..$ mat:List of 5\n",
      "  .. ..$ : int [1:177293] 0 1 1 1 1 1 1 0 1 1 ...\n",
      "  .. ..$ : int [1:185037] 0 0 1 1 0 0 0 1 1 0 ...\n",
      "  .. ..$ : int [1:152385] 1 1 0 1 0 1 0 1 1 0 ...\n",
      "  .. ..$ : int [1:145964] 0 0 1 1 1 0 1 1 0 0 ...\n",
      "  .. ..$ : int [1:139435] 1 1 0 0 0 1 1 1 1 0 ...\n"
     ]
    }
   ],
   "source": [
    "haplos <- recomb_haplo_inds(inds, haplo)\n",
    "str(haplos[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "485634f0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "## Extract Phased Genotypes\n",
    "While the `simfam` documentation includes a function to convert the haplotypes into a genotype matrix, we **dont want to use it** because it removes the phasing information (i.e. we won't know which alleles were inherited from which parent). To remedy this, we will roll our own function to extract the information. This function collapses the two parental haplotypes into a single genotype and converts the vector of vectors into a single vector of all the genotypes for that one individual (using `unlist()`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e3df59d9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "extract_phased_geno <- function(hap){\n",
    "    sapply(seq_along(hap$pat), function(x){paste(hap$pat[[x]], hap$mat[[x]], sep = \"|\")}) |>\n",
    "    unlist()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4c316a10",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>father</th><th scope=col>mother</th><th scope=col>offspring_1</th><th scope=col>offspring_2</th><th scope=col>offspring_3</th><th scope=col>offspring_4</th><th scope=col>offspring_5</th><th scope=col>offspring_6</th><th scope=col>offspring_7</th><th scope=col>offspring_8</th><th scope=col>offspring_9</th><th scope=col>offspring_10</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1|0</td><td>1|1</td><td>1|1</td><td>0|1</td><td>1|1</td><td>1|1</td><td>0|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>0|1</td><td>1|1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1|1</td><td>0|1</td><td>1|0</td><td>1|1</td><td>1|0</td><td>1|0</td><td>1|0</td><td>1|1</td><td>1|1</td><td>1|0</td><td>1|1</td><td>1|1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>0|1</td><td>1|0</td><td>0|1</td><td>1|0</td><td>0|1</td><td>0|1</td><td>1|1</td><td>0|0</td><td>0|0</td><td>0|1</td><td>1|0</td><td>0|0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>0|1</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>0|1</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & father & mother & offspring\\_1 & offspring\\_2 & offspring\\_3 & offspring\\_4 & offspring\\_5 & offspring\\_6 & offspring\\_7 & offspring\\_8 & offspring\\_9 & offspring\\_10\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1\\textbar{}0 & 1\\textbar{}1 & 1\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}1\\\\\n",
       "\t2 & 1\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}0 & 1\\textbar{}1 & 1\\textbar{}0 & 1\\textbar{}0 & 1\\textbar{}0 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}0 & 1\\textbar{}1 & 1\\textbar{}1\\\\\n",
       "\t3 & 0\\textbar{}1 & 1\\textbar{}0 & 0\\textbar{}1 & 1\\textbar{}0 & 0\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}1 & 0\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}1 & 1\\textbar{}0 & 0\\textbar{}0\\\\\n",
       "\t4 & 0\\textbar{}1 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0\\\\\n",
       "\t5 & 0\\textbar{}1 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0\\\\\n",
       "\t6 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | father &lt;chr&gt; | mother &lt;chr&gt; | offspring_1 &lt;chr&gt; | offspring_2 &lt;chr&gt; | offspring_3 &lt;chr&gt; | offspring_4 &lt;chr&gt; | offspring_5 &lt;chr&gt; | offspring_6 &lt;chr&gt; | offspring_7 &lt;chr&gt; | offspring_8 &lt;chr&gt; | offspring_9 &lt;chr&gt; | offspring_10 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1|0 | 1|1 | 1|1 | 0|1 | 1|1 | 1|1 | 0|1 | 1|1 | 1|1 | 1|1 | 0|1 | 1|1 |\n",
       "| 2 | 1|1 | 0|1 | 1|0 | 1|1 | 1|0 | 1|0 | 1|0 | 1|1 | 1|1 | 1|0 | 1|1 | 1|1 |\n",
       "| 3 | 0|1 | 1|0 | 0|1 | 1|0 | 0|1 | 0|1 | 1|1 | 0|0 | 0|0 | 0|1 | 1|0 | 0|0 |\n",
       "| 4 | 0|1 | 0|0 | 0|0 | 1|0 | 0|0 | 0|0 | 1|0 | 0|0 | 0|0 | 0|0 | 1|0 | 0|0 |\n",
       "| 5 | 0|1 | 0|0 | 0|0 | 1|0 | 0|0 | 0|0 | 1|0 | 0|0 | 0|0 | 0|0 | 1|0 | 0|0 |\n",
       "| 6 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 |\n",
       "\n"
      ],
      "text/plain": [
       "  father mother offspring_1 offspring_2 offspring_3 offspring_4 offspring_5\n",
       "1 1|0    1|1    1|1         0|1         1|1         1|1         0|1        \n",
       "2 1|1    0|1    1|0         1|1         1|0         1|0         1|0        \n",
       "3 0|1    1|0    0|1         1|0         0|1         0|1         1|1        \n",
       "4 0|1    0|0    0|0         1|0         0|0         0|0         1|0        \n",
       "5 0|1    0|0    0|0         1|0         0|0         0|0         1|0        \n",
       "6 1|1    1|1    1|1         1|1         1|1         1|1         1|1        \n",
       "  offspring_6 offspring_7 offspring_8 offspring_9 offspring_10\n",
       "1 1|1         1|1         1|1         0|1         1|1         \n",
       "2 1|1         1|1         1|0         1|1         1|1         \n",
       "3 0|0         0|0         0|1         1|0         0|0         \n",
       "4 0|0         0|0         0|0         1|0         0|0         \n",
       "5 0|0         0|0         0|0         1|0         0|0         \n",
       "6 1|1         1|1         1|1         1|1         1|1         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genotype_df <- lapply(names(haplos), function(x){extract_phased_geno(haplos[[x]])})\n",
    "names(genotype_df) <- names(haplos)\n",
    "genotype_df <- as.data.frame(genotype_df)\n",
    "head(genotype_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "468e6cc6",
   "metadata": {},
   "source": [
    "## Adding the records\n",
    "Most of the heavy lifting was already done during [the reference genome processing](../reference_assembly/setup_reference.ipynb) and the VCF file header was [previously setup](./setup_vcf.ipynb). The former output a file that already contains the first few columns of a VCF record: `CHROM`, `POS`, `ID`, `REF`, `ALT`, and we created our genotypes here. All that's left is to add the missing `QUAL`, `FILTER`, `INFO` (values don't matter), and `FORMAT` columns to `snp_pos`, then add the genotype columns and we have ourselves most of a VCF!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ea904b77",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 21</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>chrom</th><th scope=col>pos</th><th scope=col>id</th><th scope=col>ref</th><th scope=col>alt</th><th scope=col>qual</th><th scope=col>filter</th><th scope=col>info</th><th scope=col>format</th><th scope=col>father</th><th scope=col>⋯</th><th scope=col>offspring_1</th><th scope=col>offspring_2</th><th scope=col>offspring_3</th><th scope=col>offspring_4</th><th scope=col>offspring_5</th><th scope=col>offspring_6</th><th scope=col>offspring_7</th><th scope=col>offspring_8</th><th scope=col>offspring_9</th><th scope=col>offspring_10</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>NC_000001.11</td><td>10870</td><td>1</td><td>G</td><td>C</td><td>.</td><td>PASS</td><td>AC=3;AN=14</td><td>GT</td><td>1|0</td><td>⋯</td><td>1|1</td><td>0|1</td><td>1|1</td><td>1|1</td><td>0|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>0|1</td><td>1|1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>NC_000001.11</td><td>11245</td><td>2</td><td>G</td><td>A</td><td>.</td><td>PASS</td><td>AC=3;AN=14</td><td>GT</td><td>1|1</td><td>⋯</td><td>1|0</td><td>1|1</td><td>1|0</td><td>1|0</td><td>1|0</td><td>1|1</td><td>1|1</td><td>1|0</td><td>1|1</td><td>1|1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>NC_000001.11</td><td>13626</td><td>3</td><td>G</td><td>T</td><td>.</td><td>PASS</td><td>AC=3;AN=14</td><td>GT</td><td>0|1</td><td>⋯</td><td>0|1</td><td>1|0</td><td>0|1</td><td>0|1</td><td>1|1</td><td>0|0</td><td>0|0</td><td>0|1</td><td>1|0</td><td>0|0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>NC_000001.11</td><td>15195</td><td>4</td><td>T</td><td>G</td><td>.</td><td>PASS</td><td>AC=3;AN=14</td><td>GT</td><td>0|1</td><td>⋯</td><td>0|0</td><td>1|0</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NC_000001.11</td><td>15981</td><td>5</td><td>A</td><td>C</td><td>.</td><td>PASS</td><td>AC=3;AN=14</td><td>GT</td><td>0|1</td><td>⋯</td><td>0|0</td><td>1|0</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td><td>0|0</td><td>0|0</td><td>1|0</td><td>0|0</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>NC_000001.11</td><td>17245</td><td>6</td><td>T</td><td>A</td><td>.</td><td>PASS</td><td>AC=3;AN=14</td><td>GT</td><td>1|1</td><td>⋯</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td><td>1|1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 21\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & chrom & pos & id & ref & alt & qual & filter & info & format & father & ⋯ & offspring\\_1 & offspring\\_2 & offspring\\_3 & offspring\\_4 & offspring\\_5 & offspring\\_6 & offspring\\_7 & offspring\\_8 & offspring\\_9 & offspring\\_10\\\\\n",
       "  & <chr> & <int> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & NC\\_000001.11 & 10870 & 1 & G & C & . & PASS & AC=3;AN=14 & GT & 1\\textbar{}0 & ⋯ & 1\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}1\\\\\n",
       "\t2 & NC\\_000001.11 & 11245 & 2 & G & A & . & PASS & AC=3;AN=14 & GT & 1\\textbar{}1 & ⋯ & 1\\textbar{}0 & 1\\textbar{}1 & 1\\textbar{}0 & 1\\textbar{}0 & 1\\textbar{}0 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}0 & 1\\textbar{}1 & 1\\textbar{}1\\\\\n",
       "\t3 & NC\\_000001.11 & 13626 & 3 & G & T & . & PASS & AC=3;AN=14 & GT & 0\\textbar{}1 & ⋯ & 0\\textbar{}1 & 1\\textbar{}0 & 0\\textbar{}1 & 0\\textbar{}1 & 1\\textbar{}1 & 0\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}1 & 1\\textbar{}0 & 0\\textbar{}0\\\\\n",
       "\t4 & NC\\_000001.11 & 15195 & 4 & T & G & . & PASS & AC=3;AN=14 & GT & 0\\textbar{}1 & ⋯ & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0\\\\\n",
       "\t5 & NC\\_000001.11 & 15981 & 5 & A & C & . & PASS & AC=3;AN=14 & GT & 0\\textbar{}1 & ⋯ & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 0\\textbar{}0 & 1\\textbar{}0 & 0\\textbar{}0\\\\\n",
       "\t6 & NC\\_000001.11 & 17245 & 6 & T & A & . & PASS & AC=3;AN=14 & GT & 1\\textbar{}1 & ⋯ & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1 & 1\\textbar{}1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 21\n",
       "\n",
       "| <!--/--> | chrom &lt;chr&gt; | pos &lt;int&gt; | id &lt;int&gt; | ref &lt;chr&gt; | alt &lt;chr&gt; | qual &lt;chr&gt; | filter &lt;chr&gt; | info &lt;chr&gt; | format &lt;chr&gt; | father &lt;chr&gt; | ⋯ ⋯ | offspring_1 &lt;chr&gt; | offspring_2 &lt;chr&gt; | offspring_3 &lt;chr&gt; | offspring_4 &lt;chr&gt; | offspring_5 &lt;chr&gt; | offspring_6 &lt;chr&gt; | offspring_7 &lt;chr&gt; | offspring_8 &lt;chr&gt; | offspring_9 &lt;chr&gt; | offspring_10 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | NC_000001.11 | 10870 | 1 | G | C | . | PASS | AC=3;AN=14 | GT | 1|0 | ⋯ | 1|1 | 0|1 | 1|1 | 1|1 | 0|1 | 1|1 | 1|1 | 1|1 | 0|1 | 1|1 |\n",
       "| 2 | NC_000001.11 | 11245 | 2 | G | A | . | PASS | AC=3;AN=14 | GT | 1|1 | ⋯ | 1|0 | 1|1 | 1|0 | 1|0 | 1|0 | 1|1 | 1|1 | 1|0 | 1|1 | 1|1 |\n",
       "| 3 | NC_000001.11 | 13626 | 3 | G | T | . | PASS | AC=3;AN=14 | GT | 0|1 | ⋯ | 0|1 | 1|0 | 0|1 | 0|1 | 1|1 | 0|0 | 0|0 | 0|1 | 1|0 | 0|0 |\n",
       "| 4 | NC_000001.11 | 15195 | 4 | T | G | . | PASS | AC=3;AN=14 | GT | 0|1 | ⋯ | 0|0 | 1|0 | 0|0 | 0|0 | 1|0 | 0|0 | 0|0 | 0|0 | 1|0 | 0|0 |\n",
       "| 5 | NC_000001.11 | 15981 | 5 | A | C | . | PASS | AC=3;AN=14 | GT | 0|1 | ⋯ | 0|0 | 1|0 | 0|0 | 0|0 | 1|0 | 0|0 | 0|0 | 0|0 | 1|0 | 0|0 |\n",
       "| 6 | NC_000001.11 | 17245 | 6 | T | A | . | PASS | AC=3;AN=14 | GT | 1|1 | ⋯ | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 | 1|1 |\n",
       "\n"
      ],
      "text/plain": [
       "  chrom        pos   id ref alt qual filter info       format father ⋯\n",
       "1 NC_000001.11 10870 1  G   C   .    PASS   AC=3;AN=14 GT     1|0    ⋯\n",
       "2 NC_000001.11 11245 2  G   A   .    PASS   AC=3;AN=14 GT     1|1    ⋯\n",
       "3 NC_000001.11 13626 3  G   T   .    PASS   AC=3;AN=14 GT     0|1    ⋯\n",
       "4 NC_000001.11 15195 4  T   G   .    PASS   AC=3;AN=14 GT     0|1    ⋯\n",
       "5 NC_000001.11 15981 5  A   C   .    PASS   AC=3;AN=14 GT     0|1    ⋯\n",
       "6 NC_000001.11 17245 6  T   A   .    PASS   AC=3;AN=14 GT     1|1    ⋯\n",
       "  offspring_1 offspring_2 offspring_3 offspring_4 offspring_5 offspring_6\n",
       "1 1|1         0|1         1|1         1|1         0|1         1|1        \n",
       "2 1|0         1|1         1|0         1|0         1|0         1|1        \n",
       "3 0|1         1|0         0|1         0|1         1|1         0|0        \n",
       "4 0|0         1|0         0|0         0|0         1|0         0|0        \n",
       "5 0|0         1|0         0|0         0|0         1|0         0|0        \n",
       "6 1|1         1|1         1|1         1|1         1|1         1|1        \n",
       "  offspring_7 offspring_8 offspring_9 offspring_10\n",
       "1 1|1         1|1         0|1         1|1         \n",
       "2 1|1         1|0         1|1         1|1         \n",
       "3 0|0         0|1         1|0         0|0         \n",
       "4 0|0         0|0         1|0         0|0         \n",
       "5 0|0         0|0         1|0         0|0         \n",
       "6 1|1         1|1         1|1         1|1         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_pos$qual <- \".\"\n",
    "snp_pos$filter <- \"PASS\"\n",
    "snp_pos$info <- \"AC=3;AN=14\"\n",
    "snp_pos$format <- \"GT\"\n",
    "\n",
    "genotypes <- cbind(snp_pos, genotype_df)\n",
    "head(genotypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19f409cc",
   "metadata": {},
   "source": [
    "What's left is to append this, sans column names, to the `pedigree.vcf` file we created [here](./generate_vcf.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0a7bdb45",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "write.table(\n",
    "    genotypes,\n",
    "    file = \"pedigree.vcf\",\n",
    "    col.names = F,\n",
    "    row.names = F,\n",
    "    sep = \"\\t\",\n",
    "    quote = F,\n",
    "    append = T\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f24cdc7",
   "metadata": {},
   "source": [
    "And for the sake of using disk space sensibly, we'll convert the VCF to a BCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2f3ed5e5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "system(\"bcftools view -Ob -o pedigree.bcf pedigree.vcf\")\n",
    "file.remove(\"pedigree.vcf\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
