{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d111d0af",
   "metadata": {},
   "source": [
    "# Setup the reference genomes\n",
    "We (unfortunately) need two reference genomes for this project:\n",
    "\n",
    "```{topic} Phasing\n",
    "Since phasing is a smaller affair, we can use the first 5 autosomes in the human genome. For read simulation purposes, we will need to convert N nucleotides in the assembly to random ATCG nucleotides. After reads are simulated using that, we will then perform all subsequent analyses using the 5-autosome assembly sans N conversions.\n",
    "```\n",
    "\n",
    "```{topic} Assembly\n",
    "Assembly needs to be done on the entire human genome, and like phasing, we need to initially convert Ns into ATCG nucleotides for read-simulation purposes. The post-processesing and benchmarking will be in relation to the unmodified reference.\n",
    "```\n",
    "\n",
    "This will give us a total of 4 reference variants:\n",
    "| | Full | 5-autosome |\n",
    "|:---|:---:|:---:|\n",
    "| **Unmodified** | Full unmodified | 5-autosome unmodified |\n",
    "| **N-converted** | Full N-converted | 5-autosome N-converted |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1631e349",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import pysam\n",
    "import re\n",
    "import random\n",
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9bd87a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(\"../reference_assembly\", exist_ok=True)\n",
    "assembly_original = \"GRCh38.p14.fasta.gz\"\n",
    "assembly_noN = \"GRCh38.p14.noN.fa\"\n",
    "assembly_trunc = \"GRCh38.p14.5A.fa\"\n",
    "assembly_trunc_noN = \"GRCh38.p14.5A.noN.fa\"\n",
    "random.seed(6969)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8409b489",
   "metadata": {},
   "source": [
    "## Download the human genome `GRCh38.p14`\n",
    "We need the human genome that was used for the pedigree genotypes (contig names, lengths, etc.) We'll download the entire thing, then subset it to keep only the first 5 autosomes that we are interested in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebfd2b15",
   "metadata": {},
   "outputs": [],
   "source": [
    "GRCh38 = \"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_genomic.fna.gz\"\n",
    "r = requests.get(GRCh38)\n",
    "open(assembly_original, 'wb').write(r.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "292f3698",
   "metadata": {},
   "source": [
    "Here we are subsetting `GRCh38` for the 5 autosomes we are interested in.\n",
    "However, we also need to do one more thing. If we use chromosomes that have long stretches of N nucleotides, we will end up being unable to simulate reads from those genomic regions. Ultimately, using an actual human reference genome means it's incomplete, but it does have natural things in it that we can't otherwise create with a random ATGC simulation (or other simulators I've found), which includes:\n",
    "- centromeric regions\n",
    "- repeat regions scattered all over\n",
    "- non-random association between adjacent nucleotides\n",
    "\n",
    "Unlike the Mimick situation, we aren't building a new simulator here ðŸ˜…, instead we are just going to **replace N nucleotides with random ATCG bases** to create the input haploid genome for the pedigree/phase linked read simulations. For the assembly benchmark, we'll later add some mutations to create a diploid. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "06c4e3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_chrom = {\"NC_000001.11\", \"NC_000002.12\", \"NC_000003.12\", \"NC_000004.12\", \"NC_000005.10\"}\n",
    "with (\n",
    "    gzip.open(assembly_original, 'r') as fain,\n",
    "    open(assembly_trunc, 'w') as faout,\n",
    "    open(assembly_trunc_noN, 'w') as faoutN,\n",
    "    open(assembly_noN, 'w') as faoutorigN\n",
    "    ):\n",
    "    write = False\n",
    "    _name = \"\"\n",
    "    for i in fain:\n",
    "        if i.startswith(b\">\"):\n",
    "            _name = i.decode(\"utf8\").split()[0].lstrip(\">\")\n",
    "            faoutorigN.write(f\">{_name}\\n\")\n",
    "            if _name in target_chrom:\n",
    "                faout.write(f\">{_name}\\n\")\n",
    "                faoutN.write(f\">{_name}\\n\")\n",
    "                _write = True\n",
    "            else:\n",
    "                _write = False\n",
    "        else:\n",
    "            _i = i.decode(\"utf-8\")\n",
    "            seq = ''.join(random.choice(\"ATCG\") if nuc == 'N' else nuc for nuc in _i)\n",
    "            faoutorigN.write(seq)\n",
    "            if _write:\n",
    "                faoutN.write(seq)\n",
    "                faout.write(_i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90f86f24",
   "metadata": {},
   "source": [
    "Below is a small sanity check to ensure that the N bases are preserved in `assembly_trunc` and replaced in `assembly_trunc_noN`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0148a2b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "truncated\n",
      ">NC_000001.11\n",
      " NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "\n",
      "truncated_noN\n",
      ">NC_000001.11\n",
      " CTGGAATCCGTCGCGAGTCGACCAACACTGGTCGCGTACAGCGGTAAGGCAGAGACACCATCCGCCACGCCGTAGGTAGG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for k,v in {\"truncated\": assembly_trunc, \"truncated_noN\": assembly_trunc_noN}.items():\n",
    "    with open(v, 'r') as fa:\n",
    "        print(k)\n",
    "        print(fa.readline(), fa.readline())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40360787",
   "metadata": {},
   "source": [
    "Finally, index the genomes for random access later. This also conveniently stores the contig lengths in the `.fai` files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f2add779",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in [assembly_trunc, assembly_trunc_noN, assembly_noN]:\n",
    "    os.system(f\"bgzip {i}\")\n",
    "    os.system(f\"samtools faidx {i}.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02b72bd6",
   "metadata": {},
   "source": [
    "## Identify non-N positions\n",
    ":::{caution} Slow AF\n",
    "The ref/alt part takes **forever** (5-6 min).\n",
    ":::\n",
    "\n",
    "This is more of an annoying detail, but in order to simulate pedigree genotypes _consistently_ (and functionally), we need to make sure which positions are viable candidates for SNPs by finding the positions of those that aren't `N`. Since there are a LOT, we'll also randomly sample the valid positions in the loop and save it as a file that will be consumed by R in the [pedigree creation](../simulate_genotypes/simulate_pedigree.ipynb). Since we're already parsing through the positions, it would also be quite convenient to get the nucleotide at the position (the `REF`) and randomly choose an `ALT`. This will make the [VCF generation](../simulate_genotypes/generate_vcf.ipynb) much easier (and less redundant)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0dfdbeb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_pos(name: str, pos: int, id: int, nuc: str) -> str:\n",
    "    ref = nuc.upper()\n",
    "    alt = random.choice(\"ATCG\".replace(ref,\"\"))\n",
    "    return f\"{name}\\t{pos}\\t{id}\\t{ref}\\t{alt}\\n\"\n",
    "\n",
    "with pysam.FastxFile(assembly_trunc + \".gz\") as fa:\n",
    "    with open(\"GRCh38.p14.snp.positions\", 'w') as positions:\n",
    "        for i in fa:\n",
    "            ID = 0\n",
    "            non_N = [match.start() + 1 for match in re.finditer(r'[^Nn]', i.sequence)]\n",
    "            for j in sorted(random.sample(non_N, round(len(non_N) / 1300))):\n",
    "                ID += 1\n",
    "                positions.write(process_pos(i.name, j, ID, i.sequence[j]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec34ec3c",
   "metadata": {},
   "source": [
    "And as a quick sanity check, inspect the first 100 bytes of the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "03cbb273",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_000001.11\t10870\t1\tG\tC\n",
      "NC_000001.11\t11245\t2\tG\tA\n",
      "NC_000001.11\t13626\t3\tG\tT\n",
      "NC_000001.11\t15195\t4\tT\tG\n",
      "NC_000001.11\t15981\t5\tA\tC\n",
      "NC_000001.11\t17245\t6\tT\tA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    open(\"GRCh38.p14.snp.positions\", 'r').read(150)\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
