{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d111d0af",
   "metadata": {},
   "source": [
    "# Setup the reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1631e349",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import random\n",
    "import requests\n",
    "\n",
    "random.seed(6969)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9bd87a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(\"../reference_assembly\", exist_ok=True)\n",
    "assembly_original = \"GRCh38.p14.fasta.gz\"\n",
    "assembly_trunc = \"GRCh38.p14.5A.fa\"\n",
    "assembly_trunc_noN = \"GRCh38.p14.5A.noN.fa\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8409b489",
   "metadata": {},
   "source": [
    "## Download the human genome `GRCh38.p14`\n",
    "We need the human genome that was used for the pedigree genotypes (contig names, lengths, etc.) We'll download the entire thing, then subset it to keep only the first 5 autosomes that we are interested in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebfd2b15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "972898531"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRCh38 = \"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_genomic.fna.gz\"\n",
    "r = requests.get(GRCh38)\n",
    "open(assembly_original, 'wb').write(r.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "292f3698",
   "metadata": {},
   "source": [
    "Here we are subsetting `GRCh38` for the 5 autosomes we are interested in.\n",
    "However, we also need to do one more thing. If we use chromosomes that have long stretches of N nucleotides, we will end up being unable to simulate reads from those genomic regions. Ultimately, using an actual human reference genome means it's incomplete, but it does have natural things in it that we can't otherwise create with a random ATGC simulation (or other simulators I've found), which includes:\n",
    "- centromeric regions\n",
    "- repeat regions scattered all over\n",
    "- non-random association between adjacent nucleotides\n",
    "\n",
    "Unlike the Mimick situation, we aren't building a new simulator here ðŸ˜…, instead we are just going to **replace N nucleotides with random ATCG bases** to create the input haploid genome for the pedigree/phase linked read simulations. For the assembly benchmark, we'll later add some mutations to create a diploid. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "06c4e3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_chrom = {\"NC_000001.11\", \"NC_000002.12\", \"NC_000003.12\", \"NC_000004.12\", \"NC_000005.10\"}\n",
    "with (\n",
    "    gzip.open(assembly_original, 'r') as fain,\n",
    "    open(assembly_trunc, 'w') as faout,\n",
    "    open(assembly_trunc_noN, 'w') as faoutN,\n",
    "    ):\n",
    "    write = False\n",
    "    _name = \"\"\n",
    "    for i in fain:\n",
    "        if i.startswith(b\">\"):\n",
    "            _name = i.decode(\"utf8\").split()[0].lstrip(\">\")\n",
    "            if _name in target_chrom:\n",
    "                faout.write(f\">{_name}\\n\")\n",
    "                faoutN.write(f\">{_name}\\n\")\n",
    "                _write = True\n",
    "            else:\n",
    "                _write = False\n",
    "        else:\n",
    "            if _write:\n",
    "                faoutN.write(\n",
    "                    ''.join(random.choice(\"ATCG\") if nuc == 'N' else nuc for nuc in i.decode(\"utf-8\"))\n",
    "                )\n",
    "                faout.write(i.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90f86f24",
   "metadata": {},
   "source": [
    "Below is a small sanity check to ensure that the N bases are preserved in `assembly_trunc` and replaced in `assembly_trunc_noN`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0148a2b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "truncated\n",
      ">NC_000001.11\n",
      "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "\n",
      "truncated_noN\n",
      ">NC_000001.11\n",
      "AATTGATCTCCCGAGAGCCCTGACCTGCTAGCGCGTCCCAGGTCGGTAATAGTAGGAACGCGCGATATTTGAAGCACAAG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with (\n",
    "    open(assembly_trunc, 'r') as fa,\n",
    "    open(assembly_trunc_noN, 'r') as faN,\n",
    "    ):\n",
    "    for k,v in {\"truncated\": fa, \"truncated_noN\": faN}.items():\n",
    "        print(k)\n",
    "        print(v.readline() , end = \"\")\n",
    "        print(v.readline())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40360787",
   "metadata": {},
   "source": [
    "## Housekeeping\n",
    "Finally, index the genomes for random access later. This also conveniently stores the contig lengths in the `.fai` files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f2add779",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system(f\"bgzip {assembly_trunc}\")\n",
    "os.system(f\"samtools faidx {assembly_trunc}.gz\")\n",
    "\n",
    "os.system(f\"bgzip {assembly_trunc_noN}\")\n",
    "os.system(f\"samtools faidx {assembly_trunc_noN}.gz\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
